gneiss/sort.py

```diff
@@ -0,0 +1,109 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, gneiss development team.
+#
+# Distributed under the terms of the GPLv3 License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+import numpy as np
+import pandas as pd
+from functools import partial
+from gneiss.util import match
+
+
+def mean_niche_estimator(abundances, gradient):
+    """ Estimates the mean niche of an organism.
+
+    Calculates the mean niche of an organism along a gradient.
+    This is done by calculating the expected value of an organism
+    across the gradient.
+
+    Specifically, this module calculates the following
+
+    .. math::
+        E[g | x] =
+        \frac{1}{\sum\limits_{j=1}^N x_j} \sum\limits_{i=1}^N x_i g_i
+
+    Where :math:`N` is the number of samples, :math:`x_i` is the proportion of
+    species :math:`x` in sample :math:`i`, :math:`g_i` is the gradient value
+    at sample `i`.
+
+    Parameters
+    ----------
+    abundances : pd.Series, np.float
+        Vector of fraction abundances of an organism over
+        a list of samples.
+    gradient : pd.Series, np.float
+        Vector of numerical gradient values.
+
+    Returns
+    -------
+    np.float :
+        The mean gradient that the organism lives in.
+
+    Raises
+    ------
+    ValueError:
+        If the length of `abundances` is not the same length as `gradient`.
+    ValueError:
+        If the length of `gradient` contains nans.
+    """
+    len_abundances = len(abundances)
+    len_gradient = len(gradient)
+    if len_abundances != len_gradient:
+        raise ValueError("Length of `abundances` (%d) doesn't match the length"
+                         " of the `gradient` (%d)" % (len_abundances,
+                                                      len_gradient))
+    if np.any(pd.isnull(gradient)):
+        raise ValueError("`gradient` cannot have any nans.")
+
+    # normalizes the proportions of the organism across all of the
+    # samples to add to 1.
+    v = abundances / abundances.sum()
+    m = np.dot(gradient, v)
+    return m
+
+
+def niche_sort(table, gradient, niche_estimator=mean_niche_estimator):
```

Comment:
It does but perhaps it will be more helpful to give an error with the available methods vs. a simple not defined:
```python
In [1]: def estimator(a, b):
   ...:     return a+b
   ...: 

In [2]: def command(a, b, method=estimator):
   ...:     return method(a, b)
   ...: 

In [3]: command(1, 2)
Out[3]: 3

In [4]: command(1, 2, method=bla)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-4-b39186e5fe69> in <module>()
----> 1 command(1, 2, method=bla)

NameError: name 'bla' is not defined
```

However, I can see that this is way beyond the scope of this PR, perhaps adding an issue will be OK for now, what do you think?